File: tools/cabana/pycabana/dialogs/streamselector.py
"""StreamSelector - dialog for selecting and opening CAN data streams."""

from typing import Optional

from PySide6.QtCore import Qt
from PySide6.QtWidgets import (
  QDialog,
  QWidget,
  QVBoxLayout,
  QHBoxLayout,
  QGridLayout,
  QFormLayout,
  QTabWidget,
  QLineEdit,
  QLabel,
  QPushButton,
  QDialogButtonBox,
  QFileDialog,
  QFrame,
  QComboBox,
  QCheckBox,
  QRadioButton,
  QButtonGroup,
)
from PySide6.QtGui import QRegularExpressionValidator
from PySide6.QtCore import QRegularExpression

from openpilot.tools.cabana.pycabana.streams.abstract import AbstractStream
from openpilot.tools.cabana.pycabana.streams.replay import ReplayStream
from openpilot.tools.cabana.pycabana.settings import Settings


class AbstractOpenStreamWidget(QWidget):
  """Base class for stream source widgets.

  Subclasses must implement open() to return a configured stream.
  """

  def __init__(self, parent: Optional[QWidget] = None):
    super().__init__(parent)

  def open(self) -> Optional[AbstractStream]:
    """Open and return a configured stream, or None on failure."""
    raise NotImplementedError("Subclasses must implement open()")


class OpenReplayWidget(AbstractOpenStreamWidget):
  """Widget for opening replay streams from routes."""

  def __init__(self, parent: Optional[QWidget] = None):
    super().__init__(parent)

    settings = Settings()

    layout = QGridLayout(self)

    # Route input
    layout.addWidget(QLabel("Route"), 0, 0)
    self.route_edit = QLineEdit(self)
    self.route_edit.setPlaceholderText("Enter route name or browse for local/remote route")
    layout.addWidget(self.route_edit, 0, 1)

    # Remote route button
    browse_remote_btn = QPushButton("Remote route...", self)
    layout.addWidget(browse_remote_btn, 0, 2)
    browse_remote_btn.setEnabled(False)  # Not implemented yet

    # Local route button
    browse_local_btn = QPushButton("Local route...", self)
    layout.addWidget(browse_local_btn, 0, 3)

    # Camera options
    camera_layout = QHBoxLayout()
    self.road_camera = QCheckBox("Road camera", self)
    self.driver_camera = QCheckBox("Driver camera", self)
    self.wide_camera = QCheckBox("Wide road camera", self)

    self.road_camera.setChecked(True)

    camera_layout.addWidget(self.road_camera)
    camera_layout.addWidget(self.driver_camera)
    camera_layout.addWidget(self.wide_camera)
    camera_layout.addStretch(1)
    layout.addLayout(camera_layout, 1, 1)

    self.setMinimumWidth(550)

    # Connect signals
    browse_local_btn.clicked.connect(self._browse_local_route)
    browse_remote_btn.clicked.connect(self._browse_remote_route)

  def _browse_local_route(self) -> None:
    """Browse for a local route directory."""
    settings = Settings()
    directory = QFileDialog.getExistingDirectory(
      self,
      "Open Local Route",
      settings.last_route_dir
    )
    if directory:
      self.route_edit.setText(directory)
      settings.last_route_dir = directory
      settings.save()

  def _browse_remote_route(self) -> None:
    """Browse for a remote route (not implemented yet)."""
    # TODO: Implement remote route browser dialog
    pass

  def open(self) -> Optional[AbstractStream]:
    """Open a replay stream from the specified route."""
    route = self.route_edit.text().strip()
    if not route:
      return None

    stream = ReplayStream()
    if stream.loadRoute(route):
      return stream

    return None


class OpenPandaWidget(AbstractOpenStreamWidget):
  """Widget for opening live streams from Panda devices."""

  def __init__(self, parent: Optional[QWidget] = None):
    super().__init__(parent)

    form_layout = QFormLayout(self)

    # Serial selection
    serial_layout = QHBoxLayout()
    self.serial_edit = QComboBox(self)
    serial_layout.addWidget(self.serial_edit)

    refresh_btn = QPushButton("Refresh", self)
    refresh_btn.setSizePolicy(refresh_btn.sizePolicy().horizontalPolicy(), refresh_btn.sizePolicy().verticalPolicy())
    serial_layout.addWidget(refresh_btn)

    form_layout.addRow("Serial", serial_layout)

    # Note: Panda streaming not fully implemented in Python yet
    info_label = QLabel("Panda live streaming is not yet fully implemented in pycabana.")
    info_label.setWordWrap(True)
    form_layout.addRow(info_label)

    refresh_btn.clicked.connect(self._refresh_serials)
    self.serial_edit.currentTextChanged.connect(self._on_serial_changed)

    self._refresh_serials()

  def _refresh_serials(self) -> None:
    """Refresh the list of available Panda devices."""
    self.serial_edit.clear()
    # TODO: Implement actual Panda device discovery
    self.serial_edit.addItem("(No Panda devices found)")

  def _on_serial_changed(self, serial: str) -> None:
    """Handle serial selection change."""
    pass

  def open(self) -> Optional[AbstractStream]:
    """Open a Panda stream (not yet implemented)."""
    # TODO: Implement PandaStream
    return None


class OpenSocketCanWidget(AbstractOpenStreamWidget):
  """Widget for opening SocketCAN streams."""

  def __init__(self, parent: Optional[QWidget] = None):
    super().__init__(parent)

    main_layout = QVBoxLayout(self)
    main_layout.addStretch(1)

    form_layout = QFormLayout()

    # Device selection
    device_layout = QHBoxLayout()
    self.device_edit = QComboBox(self)
    self.device_edit.setFixedWidth(300)
    device_layout.addWidget(self.device_edit)

    refresh_btn = QPushButton("Refresh", self)
    refresh_btn.setFixedWidth(100)
    device_layout.addWidget(refresh_btn)

    form_layout.addRow("Device", device_layout)
    main_layout.addLayout(form_layout)

    main_layout.addStretch(1)

    # Note: SocketCAN not fully implemented in Python yet
    info_label = QLabel("SocketCAN streaming is not yet fully implemented in pycabana.")
    info_label.setWordWrap(True)
    form_layout.addRow(info_label)

    refresh_btn.clicked.connect(self._refresh_devices)
    self.device_edit.currentTextChanged.connect(self._on_device_changed)

    self._refresh_devices()

  def _refresh_devices(self) -> None:
    """Refresh the list of available SocketCAN devices."""
    self.device_edit.clear()
    # TODO: Implement actual SocketCAN device discovery
    self.device_edit.addItem("(No SocketCAN devices found)")

  def _on_device_changed(self, device: str) -> None:
    """Handle device selection change."""
    pass

  def open(self) -> Optional[AbstractStream]:
    """Open a SocketCAN stream (not yet implemented)."""
    # TODO: Implement SocketCanStream
    return None


class OpenDeviceWidget(AbstractOpenStreamWidget):
  """Widget for opening live streams from openpilot devices."""

  def __init__(self, parent: Optional[QWidget] = None):
    super().__init__(parent)

    form_layout = QFormLayout(self)

    # Connection type selection
    self.msgq_radio = QRadioButton("MSGQ", self)
    self.zmq_radio = QRadioButton("ZMQ", self)

    # IP address input for ZMQ
    self.ip_address = QLineEdit(self)
    self.ip_address.setPlaceholderText("Enter device IP Address")

    # IP address validation
    ip_range = "(?:[0-1]?[0-9]?[0-9]|2[0-4][0-9]|25[0-5])"
    pattern = f"^{ip_range}\\.{ip_range}\\.{ip_range}\\.{ip_range}$"
    regex = QRegularExpression(pattern)
    validator = QRegularExpressionValidator(regex, self)
    self.ip_address.setValidator(validator)

    # Button group for radio buttons
    self.connection_group = QButtonGroup(self)
    self.connection_group.addButton(self.msgq_radio, 0)
    self.connection_group.addButton(self.zmq_radio, 1)

    form_layout.addRow(self.msgq_radio)
    form_layout.addRow(self.zmq_radio, self.ip_address)

    # Note: Device streaming not fully implemented in Python yet
    info_label = QLabel("Device live streaming is not yet fully implemented in pycabana.")
    info_label.setWordWrap(True)
    form_layout.addRow(info_label)

    # Default to ZMQ
    self.zmq_radio.setChecked(True)

    # Connect signals
    self.connection_group.buttonToggled.connect(self._on_connection_toggled)

  def _on_connection_toggled(self, button, checked: bool) -> None:
    """Handle connection type toggle."""
    self.ip_address.setEnabled(button == self.zmq_radio and checked)

  def open(self) -> Optional[AbstractStream]:
    """Open a device stream (not yet implemented)."""
    # TODO: Implement DeviceStream
    return None


class StreamSelector(QDialog):
  """Dialog for selecting and opening CAN data streams.

  Provides tabs for different stream sources:
  - Replay: Load data from openpilot routes
  - Panda: Connect to Panda hardware
  - SocketCAN: Connect to SocketCAN devices
  - Device: Connect to live openpilot devices
  """

  def __init__(self, parent: Optional[QWidget] = None):
    super().__init__(parent)

    self.stream_: Optional[AbstractStream] = None
    self.settings = Settings()

    self.setWindowTitle("Open stream")

    layout = QVBoxLayout(self)

    # Tab widget for different stream sources
    self.tab = QTabWidget(self)
    layout.addWidget(self.tab)

    # DBC file selection
    dbc_layout = QHBoxLayout()

    self.dbc_file = QLineEdit(self)
    self.dbc_file.setReadOnly(True)
    self.dbc_file.setPlaceholderText("Choose a dbc file to open")

    file_btn = QPushButton("Browse...", self)

    dbc_layout.addWidget(QLabel("dbc File"))
    dbc_layout.addWidget(self.dbc_file)
    dbc_layout.addWidget(file_btn)

    layout.addLayout(dbc_layout)

    # Separator line
    line = QFrame(self)
    line.setFrameStyle(QFrame.Shape.HLine | QFrame.Shadow.Sunken)
    layout.addWidget(line)

    # Dialog buttons
    self.btn_box = QDialogButtonBox(
      QDialogButtonBox.StandardButton.Open | QDialogButtonBox.StandardButton.Cancel,
      self
    )
    layout.addWidget(self.btn_box)

    # Add stream source widgets
    self._add_stream_widget(OpenReplayWidget(), "&Replay")
    self._add_stream_widget(OpenPandaWidget(), "&Panda")

    # Only add SocketCAN if available (platform dependent)
    if self._socketcan_available():
      self._add_stream_widget(OpenSocketCanWidget(), "&SocketCAN")

    self._add_stream_widget(OpenDeviceWidget(), "&Device")

    # Connect signals
    self.btn_box.rejected.connect(self.reject)
    self.btn_box.accepted.connect(self._on_open_clicked)
    file_btn.clicked.connect(self._browse_dbc_file)

  def _add_stream_widget(self, widget: AbstractOpenStreamWidget, title: str) -> None:
    """Add a stream source widget as a tab."""
    self.tab.addTab(widget, title)

  def _socketcan_available(self) -> bool:
    """Check if SocketCAN is available on this platform."""
    # SocketCAN is typically only available on Linux
    import platform
    return platform.system() == "Linux"

  def _browse_dbc_file(self) -> None:
    """Browse for a DBC file."""
    filename = QFileDialog.getOpenFileName(
      self,
      "Open File",
      self.settings.last_dir,
      "DBC (*.dbc)"
    )[0]

    if filename:
      self.dbc_file.setText(filename)
      from pathlib import Path
      self.settings.last_dir = str(Path(filename).parent)
      self.settings.save()

  def _on_open_clicked(self) -> None:
    """Handle the Open button click."""
    self.setEnabled(False)

    # Get the current widget and try to open its stream
    current_widget = self.tab.currentWidget()
    if isinstance(current_widget, AbstractOpenStreamWidget):
      self.stream_ = current_widget.open()
      if self.stream_:
        self.accept()
        return

    self.setEnabled(True)

  def dbcFile(self) -> str:
    """Get the selected DBC file path."""
    return self.dbc_file.text()

  def stream(self) -> Optional[AbstractStream]:
    """Get the opened stream."""
    return self.stream_

File: openpilot/tools/cabana/streamselector.cc
#include "tools/cabana/streamselector.h"

#include <filesystem>
#include <fstream>

#include <QFileDialog>
#include <QGridLayout>
#include <QLabel>
#include <QMessageBox>
#include <QPushButton>
#include <QRadioButton>
#include <QTimer>

#include "tools/cabana/streams/devicestream.h"
#include "tools/cabana/streams/replaystream.h"
#include "tools/cabana/routesdialog.h"
#include "tools/cabana/utils/qtutil.h"

// OpenReplayWidget

OpenReplayWidget::OpenReplayWidget(QWidget *parent) : AbstractOpenStreamWidget(parent) {
  QGridLayout *grid_layout = new QGridLayout(this);
  grid_layout->addWidget(new QLabel(tr("Route")), 0, 0);
  grid_layout->addWidget(route_edit = new QLineEdit(this), 0, 1);
  route_edit->setPlaceholderText(tr("Enter route name or browse for local/remote route"));
  auto browse_remote_btn = new QPushButton(tr("Remote route..."), this);
  grid_layout->addWidget(browse_remote_btn, 0, 2);
  auto browse_local_btn = new QPushButton(tr("Local route..."), this);
  grid_layout->addWidget(browse_local_btn, 0, 3);

  QHBoxLayout *camera_layout = new QHBoxLayout();
  for (auto c : {tr("Road camera"), tr("Driver camera"), tr("Wide road camera")})
    camera_layout->addWidget(cameras.emplace_back(new QCheckBox(c, this)));
  cameras[0]->setChecked(true);
  camera_layout->addStretch(1);
  grid_layout->addItem(camera_layout, 1, 1);

  setMinimumWidth(550);
  QObject::connect(browse_local_btn, &QPushButton::clicked, [=]() {
    QString dir = QFileDialog::getExistingDirectory(this, tr("Open Local Route"), QString::fromStdString(settings.last_route_dir));
    if (!dir.isEmpty()) {
      route_edit->setText(dir);
      settings.last_route_dir = std::filesystem::absolute(dir.toStdString()).parent_path().string();
    }
  });
  QObject::connect(browse_remote_btn, &QPushButton::clicked, [this]() {
    RoutesDialog route_dlg(this);
    if (route_dlg.exec()) {
      route_edit->setText(QString::fromStdString(route_dlg.route()));
    }
  });
}

AbstractStream *OpenReplayWidget::open() {
  QString route = route_edit->text();
  QString data_dir;
  if (int idx = route.lastIndexOf('/'); idx != -1 && util::file_exists(route.toStdString())) {
    data_dir = route.mid(0, idx + 1);
    route = route.mid(idx + 1);
  }

  bool is_valid_format = Route::parseRoute(route.toStdString()).str.size() > 0;
  if (!is_valid_format) {
    QMessageBox::warning(nullptr, tr("Warning"), tr("Invalid route format: '%1'").arg(route));
  } else {
    auto replay_stream = std::make_unique<ReplayStream>();
    Connection err = replay_stream->error.connect([](const std::string &msg) {
      QMessageBox::warning(nullptr, tr("Error"), QString::fromStdString(msg));
    });
    uint32_t flags = REPLAY_FLAG_NONE;
    if (cameras[1]->isChecked()) flags |= REPLAY_FLAG_CABIN_CAMERA;
    if (cameras[2]->isChecked()) flags |= REPLAY_FLAG_WIDE_ROAD;
    if (flags == REPLAY_FLAG_NONE && !cameras[0]->isChecked()) flags = REPLAY_FLAG_NO_VIPC;

    if (replay_stream->loadRoute(route.toStdString(), data_dir.toStdString(), flags)) {
      return replay_stream.release();
    }
  }
  return nullptr;
}

// OpenPandaWidget

static const uint32_t speeds[] = {10U, 20U, 50U, 100U, 125U, 250U, 500U, 1000U};
static const uint32_t data_speeds[] = {10U, 20U, 50U, 100U, 125U, 250U, 500U, 1000U, 2000U, 5000U};

OpenPandaWidget::OpenPandaWidget(QWidget *parent) : AbstractOpenStreamWidget(parent) {
  form_layout = new QFormLayout(this);
  if (can && dynamic_cast<PandaStream *>(can) != nullptr) {
    form_layout->addWidget(new QLabel(tr("Already connected to %1.").arg(QString::fromStdString(can->routeName()))));
    form_layout->addWidget(new QLabel("Close the current connection via [File menu -> Close Stream] before connecting to another Panda."));
    QTimer::singleShot(0, [this]() { emit enableOpenButton(false); });
    return;
  }

  QHBoxLayout *serial_layout = new QHBoxLayout();
  serial_layout->addWidget(serial_edit = new QComboBox());

  QPushButton *refresh = new QPushButton(tr("Refresh"));
  refresh->setSizePolicy(QSizePolicy::Fixed, QSizePolicy::Preferred);
  serial_layout->addWidget(refresh);
  form_layout->addRow(tr("Serial"), serial_layout);

  QObject::connect(refresh, &QPushButton::clicked, this, &OpenPandaWidget::refreshSerials);
  QObject::connect(serial_edit, &QComboBox::currentTextChanged, this, &OpenPandaWidget::buildConfigForm);

  // Populate serials
  refreshSerials();
  buildConfigForm();
}

void OpenPandaWidget::refreshSerials() {
  serial_edit->clear();
  for (auto serial : Panda::list()) {
    serial_edit->addItem(QString::fromStdString(serial));
  }
}

void OpenPandaWidget::buildConfigForm() {
  for (int i = form_layout->rowCount() - 1; i > 0; --i) {
    form_layout->removeRow(i);
  }

  QString serial = serial_edit->currentText();
  bool has_fd = false;
  bool has_panda = !serial.isEmpty();
  if (has_panda) {
    try {
      Panda panda(serial.toStdString());
      has_fd = (panda.hw_type == cereal::PandaState::PandaType::RED_PANDA) || (panda.hw_type == cereal::PandaState::PandaType::RED_PANDA_V2);
    } catch (const std::exception& e) {
      fprintf(stderr, "failed to open panda %s\n", serial.toUtf8().constData());
      has_panda = false;
    }
  }

  if (has_panda) {
    config.serial = serial.toStdString();
    config.bus_config.resize(3);
    for (int i = 0; i < config.bus_config.size(); i++) {
      QHBoxLayout *bus_layout = new QHBoxLayout;

      // CAN Speed
      bus_layout->addWidget(new QLabel(tr("CAN Speed (kbps):")));
      QComboBox *can_speed = new QComboBox;
      for (int j = 0; j < std::size(speeds); j++) {
        can_speed->addItem(QString::number(speeds[j]));

        if (data_speeds[j] == config.bus_config[i].can_speed_kbps) {
          can_speed->setCurrentIndex(j);
        }
      }
      QObject::connect(can_speed, qOverload<int>(&QComboBox::currentIndexChanged), [=](int index) {config.bus_config[i].can_speed_kbps = speeds[index];});
      bus_layout->addWidget(can_speed);

      // CAN-FD Speed
      if (has_fd) {
        QCheckBox *enable_fd = new QCheckBox("CAN-FD");
        bus_layout->addWidget(enable_fd);
        bus_layout->addWidget(new QLabel(tr("Data Speed (kbps):")));
        QComboBox *data_speed = new QComboBox;
        for (int j = 0; j < std::size(data_speeds); j++) {
          data_speed->addItem(QString::number(data_speeds[j]));

          if (data_speeds[j] == config.bus_config[i].data_speed_kbps) {
            data_speed->setCurrentIndex(j);
          }
        }

        data_speed->setEnabled(false);
        bus_layout->addWidget(data_speed);

        QObject::connect(data_speed, qOverload<int>(&QComboBox::currentIndexChanged), [=](int index) {config.bus_config[i].data_speed_kbps = data_speeds[index];});
        QObject::connect(enable_fd, &QCheckBox::stateChanged, data_speed, &QComboBox::setEnabled);
        QObject::connect(enable_fd, &QCheckBox::stateChanged, [=](int state) {config.bus_config[i].can_fd = (bool)state;});
      }

      form_layout->addRow(tr("Bus %1:").arg(i), bus_layout);
    }
  } else {
    config.serial = "";
    form_layout->addWidget(new QLabel(tr("No panda found")));
  }
}

AbstractStream *OpenPandaWidget::open() {
  try {
    return new PandaStream(config);
  } catch (std::exception &e) {
    QMessageBox::warning(nullptr, tr("Warning"), tr("Failed to connect to panda: '%1'").arg(e.what()));
    return nullptr;
  }
}

// OpenDeviceWidget

OpenDeviceWidget::OpenDeviceWidget(QWidget *parent) : AbstractOpenStreamWidget(parent) {
  QRadioButton *msgq = new QRadioButton(tr("MSGQ"));
  QRadioButton *zmq = new QRadioButton(tr("ZMQ"));
  ip_address = new QLineEdit(this);
  ip_address->setPlaceholderText(tr("Enter device Ip Address"));
  ip_address->setValidator(new IpAddressValidator(this));

  group = new QButtonGroup(this);
  group->addButton(msgq, 0);
  group->addButton(zmq, 1);

  QFormLayout *form_layout = new QFormLayout(this);
  form_layout->addRow(msgq);
  form_layout->addRow(zmq, ip_address);
  QObject::connect(group, qOverload<QAbstractButton *, bool>(&QButtonGroup::buttonToggled), [=](QAbstractButton *button, bool checked) {
    ip_address->setEnabled(button == zmq && checked);
  });
  zmq->setChecked(true);
}

AbstractStream *OpenDeviceWidget::open() {
  std::string ip = ip_address->text().isEmpty() ? "127.0.0.1" : ip_address->text().toStdString();
  bool msgq = group->checkedId() == 0;
  return new DeviceStream(msgq ? "" : ip);
}

#ifdef __linux__
// OpenSocketCanWidget

OpenSocketCanWidget::OpenSocketCanWidget(QWidget *parent) : AbstractOpenStreamWidget(parent) {
  QVBoxLayout *main_layout = new QVBoxLayout(this);
  main_layout->addStretch(1);

  QFormLayout *form_layout = new QFormLayout();

  QHBoxLayout *device_layout = new QHBoxLayout();
  device_edit = new QComboBox();
  device_edit->setFixedWidth(300);
  device_layout->addWidget(device_edit);

  QPushButton *refresh = new QPushButton(tr("Refresh"));
  refresh->setFixedWidth(100);
  device_layout->addWidget(refresh);
  form_layout->addRow(tr("Device"), device_layout);
  main_layout->addLayout(form_layout);

  main_layout->addStretch(1);

  QObject::connect(refresh, &QPushButton::clicked, this, &OpenSocketCanWidget::refreshDevices);
  QObject::connect(device_edit, &QComboBox::currentTextChanged, this, [=]{ config.device = device_edit->currentText().toStdString(); });

  // Populate devices
  refreshDevices();
}

void OpenSocketCanWidget::refreshDevices() {
  device_edit->clear();
  // Scan /sys/class/net/ for CAN interfaces (type 280 = ARPHRD_CAN)
  std::error_code ec;
  for (const auto &entry : std::filesystem::directory_iterator("/sys/class/net", ec)) {
    std::ifstream type_file(entry.path() / "type");
    int type = 0;
    if (type_file >> type && type == 280) {
      device_edit->addItem(QString::fromStdString(entry.path().filename().string()));
    }
  }
}

AbstractStream *OpenSocketCanWidget::open() {
  try {
    return new SocketCanStream(config);
  } catch (std::exception &e) {
    QMessageBox::warning(nullptr, tr("Warning"), tr("Failed to connect to SocketCAN device: '%1'").arg(e.what()));
    return nullptr;
  }
}
#endif

// StreamSelector

StreamSelector::StreamSelector(QWidget *parent) : QDialog(parent) {
  setWindowTitle(tr("Open stream"));
  QVBoxLayout *layout = new QVBoxLayout(this);
  tab = new QTabWidget(this);
  layout->addWidget(tab);

  QHBoxLayout *dbc_layout = new QHBoxLayout();
  dbc_file = new QLineEdit(this);
  dbc_file->setReadOnly(true);
  dbc_file->setPlaceholderText(tr("Choose a dbc file to open"));
  QPushButton *file_btn = new QPushButton(tr("Browse..."));
  dbc_layout->addWidget(new QLabel(tr("dbc File")));
  dbc_layout->addWidget(dbc_file);
  dbc_layout->addWidget(file_btn);
  layout->addLayout(dbc_layout);

  QFrame *line = new QFrame(this);
  line->setFrameStyle(QFrame::HLine | QFrame::Sunken);
  layout->addWidget(line);

  btn_box = new QDialogButtonBox(QDialogButtonBox::Open | QDialogButtonBox::Cancel);
  layout->addWidget(btn_box);

  addStreamWidget(new OpenReplayWidget, tr("&Replay"));
  addStreamWidget(new OpenPandaWidget, tr("&Panda"));
#ifdef __linux__
  if (SocketCanStream::available()) {
    addStreamWidget(new OpenSocketCanWidget, tr("&SocketCAN"));
  }
#endif
  addStreamWidget(new OpenDeviceWidget, tr("&Device"));

  QObject::connect(btn_box, &QDialogButtonBox::rejected, this, &QDialog::reject);
  QObject::connect(btn_box, &QDialogButtonBox::accepted, [=]() {
    setEnabled(false);
    if (stream_ = ((AbstractOpenStreamWidget *)tab->currentWidget())->open(); stream_) {
      accept();
    }
    setEnabled(true);
  });
  QObject::connect(file_btn, &QPushButton::clicked, [this]() {
    QString fn = QFileDialog::getOpenFileName(this, tr("Open File"), QString::fromStdString(settings.last_dir), "DBC (*.dbc)");
    if (!fn.isEmpty()) {
      dbc_file->setText(fn);
      settings.last_dir = std::filesystem::absolute(fn.toStdString()).parent_path().string();
    }
  });
}

void StreamSelector::addStreamWidget(AbstractOpenStreamWidget *w, const QString &title) {
  tab->addTab(w, title);
  auto open_btn = btn_box->button(QDialogButtonBox::Open);
  QObject::connect(w, &AbstractOpenStreamWidget::enableOpenButton, open_btn, &QPushButton::setEnabled);
}
